feat(codelabs): improve web-components codelab (#957)
open-wc/open-wc: packages/codelabs/src/basics/web-components.md

## Introduction

> In this codelab, you will learn about the basics of Web Cmponents and how they work.

Reviewer: Cmponents? 🙈

Author: Ugh... 🙈 

Thanks, Thomas 👍

In this codelab, you will learn about the basics of web components and how they work.

Web components are a set of low-level browser features that allow us to write modular, encapsulated and reusable HTML elements. Web components are based on web standards and work in any environment that supports basic HTML and JavaScript. This means that there is no complex setup required for you to get started.

> The features of web components align well with the way that browsers work and have always worked. This means that they are pretty low level and straightforward. For most projects, libraries or frameworks will still be used, but instead of each framework developing their own component model, they can preferably use the component technology that is baked into the browser.

Reviewer 2: ```suggestion
Web components align with the way that browsers have always worked, they are pretty low level and straightforward. For most projects you will still want to use libraries or frameworks. But instead of each framework developing their own component model, they can use the features that are already baked into the browser.
```

> Web components are quite flexible and have a multitude of possible use cases. The more prominent use case is to build a UI component using web components. This is especially powerful, for the reason that UI components can be reused in applications that are built with different technologies.

Reviewer 2: ```suggestion
Web components are quite flexible and have a multitude of possible use cases. The more prominent use case is to build reusable UI components. This is especially powerful, for the reason that UI components can be reused in applications that are built with different technologies.
```

## Templating

The second Web Components browser feature that we will look into is HTML templates. When writing web components, we usually need to do more than just setting some styles or text. We often need to render larger pieces of HTML as part of our component and update parts of it when the user interacts with it.

To do this efficiently, the browser provides us with a `<template>` element. This element allows us to define the structure of a piece of a HTML upfront, and efficiently clone it when needed. This is a lot faster than recreating the same HTML structure each time. Using and cloning templates is (intentionally) pretty low level. You can read more about the basic API [here](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/template).

<aside class="notice">

> Usually you will not want to write and use templates directly using the low level API. For writing templates, we recommend the <a href="https://github.com/Polymer/lit-html">lit-html</a> library. Next to this base library we also recommend <a href="https://github.com/Polymer/lit-element">lit-element</a>, which makes it easy to use this library in a web component. We will discuss lit-element in a following <a href="https://open-wc.org/codelabs/">codelab</a>.

Reviewer 2: ```suggestion
Usually you will not want to write and use templates directly using the low level API. For writing templates, we recommend <a href="https://github.com/Polymer/lit-html">lit-html</a> and <a href="https://github.com/Polymer/lit-element">lit-element</a>. You can learn more about these libraries in a followup <a href="https://open-wc.org/codelabs/">codelab</a>.
```

</aside>

Let's create a `<template>` inside the body of our previous excercise.

```html
<template></template>
```

Next, we'll move the inner HTML of `<cool-heading>` to the template:

```html
<template>
  <h1>Hello world!</h1>
</template>
```

> Notice that "Hello World" is not rendered anymore. As the HTML is now a child of a template, the browser interprets it as a HTML Fragment, which is not part of the page's DOM.

Reviewer 2: ```suggestion
Notice that "Hello World" is no longer rendered on the page. This is because the content of a `<template>` elements is inert. Nothing inside it is displayed, images don't get downloaded and scripts don't run.
```

Author: should we mention/explain the term DocumentFragment? You can see it in the browser tools while inspecting.
And it is inert because it’s not directly in the DOM, right? :-)

Author: should we mention/explain the term DocumentFragment? You can see it in the browser tools while inspecting.
And it is inert because it’s not directly in the DOM, right? :-)

Author: should we mention/explain the term DocumentFragment? You can see it in the browser tools while inspecting.
And it is inert because it’s not directly in the DOM, right? :-)

Reviewer 2: I'd like to not introduce terminology people don't need to know, I think it's clearer like this but open for suggestions on my suggestion :P

## Shadow DOM

The last important Web Components browser feature we will look into is Shadow DOM. Traditionally, the context of HTML and CSS have always been global. This scales pretty badly, because we constantly need to make sure that the id's of all the elements are unique and often CSS selectors can get pretty complex. This is why many front-end frameworks offer some form of encapsulation. Web Components provide us with this capability using a "Shadow DOM", this capability is now built into the browser. When adding child elements to a Shadow DOM of a component, they will not be direct children of our element, but rather they are wrapped inside of a _shadow root_.

This shadow root is a special type of DOM node which encapsulates the elements inside of it. Styles defined inside this shadow root do not leak out, and styles defined outside the shadow root do not reach in, hence encapsulation. Also, it's not possible to use a regular `querySelector()` to select elements inside or outside the shadow root. This way we can build reusable components and gives us the confidence that they will always work the same way, no matter the environment.

> Let's try this out by attaching a new, open, shadow root to our component. An open shadow root is still queryable by Javascript. This will instantiate `this.shadowRoot` as a valid property.

Reviewer 2: ```suggestion
Let's try this out by attaching a shadow root to our component, making it available as a property called `shadowRoot`.
```

I don't think we should explain open/closed here. They should just follow the example and use open :)

Author: Yes, I was not sure either. In the other codelab I saw the aside on why we should use super.connectCallback(). It seemed also quite advance. That’s why I left it here.

```js
connectedCallback() {
  const template = document.querySelector('template');
  const clone = document.importNode(template.content, true);

>   this.attachShadow({mode: 'open'});

Reviewer 2: ```suggestion
  this.attachShadow({ mode: 'open' });
```

  this.appendChild(clone);
}
```

Now, instead of adding the cloned content to the direct children, we add the content to the shadow root's children.

```js
connectedCallback() {
  const template = document.querySelector('template');
  const clone = document.importNode(template.content, true);

>   this.attachShadow({mode: 'open'});

Reviewer 2: ```suggestion
  this.attachShadow({ mode: 'open' });
```

Web Components are being used in the wild by many companies and in many projects. Some examples:

### YouTube

> YouTube is one of the most popular sites on the internet. Their main website is built with custom elements to split up the different parts of their page in components. Additionally, they are using Polymer behaviors to work with attributes and properties.

Reviewer 2: ```suggestion
YouTube is one of the most popular sites on the internet. Their main website is built with custom elements to split up the different parts of their page in components.
```

Author: No mentioning Polymer? ;-)

Reviewer 3: I think in the context its more important to mention they're using custom elements, rather than specifically mention Polymer. People new to web components may not know what polymer is and exactly how it relates to custom elements

Author: Yes in the end, these behaviors are just mixins , tight to Custom Elements (which I already mention anyway)

Reviewer 2: Yeah I really like that mention youtube, I was thinking about adding this the other day. 

I don't thinking mentioning polymer behaviors adds anything to the case here. Plus they're just using the polymer base class, right?
